File: framework/src/main/java/com/github/mrmitew/bodylog/framework/common/view/BasePresentableTextView.java
package com.github.mrmitew.bodylog.framework.common.view;

import android.content.Context;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;

import com.github.mrmitew.bodylog.adapter.common.view.BaseView;


public abstract class BasePresentableTextView<V extends BaseView<S>, S> extends AppCompatTextView implements Presentable<V, S> {
    private BasePresenterHolder<V, S> mPresenterHolder;

    public BasePresentableTextView(final Context context) {
        super(context);
        init();
    }

    public BasePresentableTextView(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public BasePresentableTextView(final Context context, final AttributeSet attrs, final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        if(!isInEditMode()) {
            setPresenterHolder();
        }
    }

    private void setPresenterHolder() {
        mPresenterHolder = injectPresenterHolder();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        mPresenterHolder.attachView(getView());
        mPresenterHolder.bindIntents();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        mPresenterHolder.detachView();
    }
}

File: framework/src/main/java/com/github/mrmitew/bodylog/framework/common/view/BasePresentableTextView.kt
package com.github.mrmitew.bodylog.framework.common.view

import android.content.Context
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet

import com.github.mrmitew.bodylog.adapter.common.UiState
import com.github.mrmitew.bodylog.adapter.common.view.BaseView
import com.github.mrmitew.bodylog.framework.common.presenter.BasePresenterHolder


abstract class BasePresentableTextView<V : BaseView<S>, S : UiState> : AppCompatTextView, Presentable<V, S> {
    private lateinit var mPresenterHolder: BasePresenterHolder<V, S>

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        if (!isInEditMode) {
            setPresenterHolder()
        }
    }

    private fun setPresenterHolder() {
        mPresenterHolder = injectPresenterHolder()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        mPresenterHolder.onAttachedToWindow(view)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mPresenterHolder.onDetachedFromWindow()
    }
}
